addons/controllers/csi/vspherecsiconfig_utils.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	csiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/csi/v1alpha1"
+)
+
+// mapVSphereCSIConfigToDataValues maps VSphereCSIConfig CR to data values
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValues(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	switch vcsiConfig.Spec.VSphereCSI.Mode {
+	case VSphereCSINonParavirtualMode:
+		return r.mapVSphereCSIConfigToDataValuesNonParavirtual(ctx, vcsiConfig, cluster)
+	case VSphereCSIParavirtualMode:
+		return r.mapVSphereCSIConfigToDataValuesParavirtual(ctx, vcsiConfig, cluster)
+	default:
+		break
+	}
+	return nil, errors.Errorf("Invalid CSI mode '%s', must either be '%s' or '%s'",
+		vcsiConfig.Spec.VSphereCSI.Mode, VSphereCSIParavirtualMode, VSphereCSINonParavirtualMode)
+}
+
+func (r *VSphereCSIConfigReconciler) mapVSphereCSIConfigToDataValuesParavirtual(ctx context.Context,
+	vcsiConfig *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (*DataValues, error) {
+	dvs := &DataValues{}
+	config := vcsiConfig.Spec.VSphereCSI.ParavirtualConfig
+
+	if config == nil {
+		return nil, errors.Errorf("ParavirtualConfig missing in Spec.VSphereCSI")
+	}
+
+	// first populate derived values
+	dvs.VSpherePVCSI = &DataValuesVSpherePVCSI{}
+
+	dvs.VSpherePVCSI.Namespace = tryParseClusterVariableString(ctx, cluster, NamespaceVarName)
+	dvs.VSpherePVCSI.ClusterName = tryParseClusterVariableString(ctx, cluster, ClusterNameVarName)
+
+	dvs.VSpherePVCSI.ClusterUID = string(cluster.UID) // only place to get UID
+
+	// default values from https://github.com/vmware-tanzu/community-edition/blob/main/addons/packages/vsphere-pv-csi/2.4.1/bundle/config/values.yaml
+	dvs.VSpherePVCSI.SupervisorMasterEndpointHostname = "supervisor.default.svc"
+	dvs.VSpherePVCSI.SupervisorMasterPort = 6443
```

Comment:
question: trying to understand why we want to set the default here ? Shouldn't we always read from VSphereCSIConfig and assign whatever value it has ?

Reply:
- the VSphereCSIConfig can be configured to have a few configurations. The controller is required to populate defaults. So no, we don't totally rely on VSphereCSIConfig object.
- true. Using a defaulting webhook is best practice.

Reply:
- these fields have been removed from vspherecsiconfig api